velox/serializers/PrestoSerializer.cpp

```diff
@@ -1757,25 +1854,45 @@ void PrestoVectorSerde::deserialize(
 
   auto pageCodecMarker = source->read<int8_t>();
   auto uncompressedSize = source->read<int32_t>();
-  // skip size in bytes
-  source->skip(4);
+  auto compressedSize = source->read<int32_t>();
   auto checksum = source->read<int64_t>();
 
   int64_t actualCheckSum = 0;
   if (isChecksumBitSet(pageCodecMarker)) {
     actualCheckSum =
-        computeChecksum(source, pageCodecMarker, numRows, uncompressedSize);
+        computeChecksum(source, pageCodecMarker, numRows, compressedSize);
   }
 
   VELOX_CHECK_EQ(
       checksum, actualCheckSum, "Received corrupted serialized page.");
 
-  // skip number of columns
-  source->skip(4);
+  VELOX_CHECK_EQ(
+      needCompression(*codec),
+      isCompressedBitSet(pageCodecMarker),
+      "Compression kind {} should align with codec marker.",
+      dwio::common::compressionKindToString(
+          dwio::common::codecTypeToCompressionKind(codec->type())));
 
   auto children = &(*result)->children();
   auto childTypes = type->as<TypeKind::ROW>().children();
-  readColumns(source, pool, childTypes, children, useLosslessTimestamp);
+  if (!needCompression(*codec)) {
+    // skip number of columns
+    source->skip(4);
```

Comment:
Shall we add a check if the number of columns match? We currently see some serialized pages corruption in meta internal test. Thanks!